File: slash/src/commands/weather.ts
import fetch from "node-fetch";
import {
  CommandContext,
  CommandOptionType,
  InteractionResponseFlags,
  SlashCommand,
  SlashCreator,
} from "slash-create";
import { EmbedBuilder } from "src/structures/EmbedBuilder";

const WEATHER_URL = `https://api.weatherapi.com/v1/forecast.json?key=${process.env.WEATHERAPI_KEY}&days=1`;

function moonPhaseEmoji(phase: string): string {
  return (phase.includes("moon") ? phase : `${phase} moon`)
    .toLowerCase()
    .replace(/\s/g, "_");
}

export default class extends SlashCommand {
  constructor(creator: SlashCreator) {
    super(creator, {
      name: "weather",
      description: "View the weather for a particular location.",
      options: [
        {
          name: "location",
          description: "The location to get weather info about.",
          type: CommandOptionType.STRING,
          required: true,
        },
        {
          name: "units",
          description: "What units you'd like the info in.",
          type: CommandOptionType.STRING,
          choices: [
            { name: "metric", value: "metric" },
            { name: "imperial", value: "imperial" },
          ],
        },
      ],
    });
  }

  async run(ctx: CommandContext) {
    const res = await fetch(`${WEATHER_URL}&q=${ctx.options.location}`);
    if (res.status === 400) {
      return ctx.send({
        content: "No weather data found for that location.",
        flags: InteractionResponseFlags.EPHEMERAL,
      });
    }

    const { location, current, forecast } = await res.json();

    // Make basic embed with all values that don't change
    const baseEmbed = new EmbedBuilder()
      .URL(`https://darksky.net/forecast/${location.lat},${location.lon}/`)
      .thumbnail(`https:${current.condition.icon}`)
      .footer(`Conditions in ${location.name}, ${location.region}`)
      .field("☀️ Sunrise", forecast.forecastday[0].astro.sunrise, true)
      .field(
        `:${moonPhaseEmoji(forecast.forecastday[0].astro.moon_phase)}: Sunset`,
        forecast.forecastday[0].astro.sunset,
        true
      )
      .blankField(true);

    if (ctx.options.units && ctx.options.units === "metric") {
      // Send data in metric if user requests it
      return ctx.send({
        embeds: baseEmbed
          .title(`${current.temp_c}°C, ${current.condition.text}`)
          .field("Wind", `${current.wind_kph} kph ${current.wind_dir}`, true)
          .field("Humidity", `${current.humidity}%`, true)
          .field("Visibility", `${current.vis_km} km`, true)
          .toJSONArr(),
      });
    } else {
      // Send data in imperial by default
      return ctx.send({
        embeds: baseEmbed
          .title(`${current.temp_f}°F, ${current.condition.text}`)
          .field("Wind", `${current.wind_mph} mph ${current.wind_dir}`, true)
          .field("Humidity", `${current.humidity}%`, true)
          .field("Visibility", `${current.vis_miles} mi`, true)
          .toJSONArr(),
      });
    }
  }
}

File: backend/commands/weather.go
package commands

import (
	"backend/types"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"os"
	"strings"
)

var weather_key = os.Getenv("WEATHER_KEY")

type WeatherData struct {
	Location struct {
		Name    string  `json:"name"`
		Region  string  `json:"region"`
		Country string  `json:"country"`
		Lat     float64 `json:"lat"`
		Lon     float64 `json:"lon"`
	} `json:"location"`
	Current struct {
		TempC     float64 `json:"temp_c"`
		TempF     float64 `json:"temp_f"`
		Condition struct {
			Text string `json:"text"`
			Icon string `json:"icon"`
		} `json:"condition"`
		WindMph  float64 `json:"wind_mph"`
		WindKph  float64 `json:"wind_kph"`
		WindDir  string  `json:"wind_dir"`
		Humidity int     `json:"humidity"`
		VisKm    float64 `json:"vis_km"`
		VisMiles float64 `json:"vis_miles"`
	} `json:"current"`
	Forecast struct {
		Forecastday []struct {
			Day struct {
				DailyChanceOfRain int `json:"daily_chance_of_rain"`
				DailyChanceOfSnow int `json:"daily_chance_of_snow"`
			} `json:"day"`
			Astro struct {
				Sunrise   string `json:"sunrise"`
				Sunset    string `json:"sunset"`
				MoonPhase string `json:"moon_phase"`
			} `json:"astro"`
		} `json:"forecastday"`
	} `json:"forecast"`
}

func weather(i *types.Interaction, data *types.ApplicationCommandInteractionData) {
	// Retrieve weather data from API
	location, _ := data.GetOption("location")
	resp, err := http.Get(fmt.Sprintf("https://api.weatherapi.com/v1/forecast.json?key=%s&days=1&q=%s", weather_key, url.QueryEscape(location.StringValue())))
	if err != nil {
		fmt.Println(err)
		i.Error("Unable to load weather data.")
		return
	}

	// Get weather data
	defer resp.Body.Close()
	var weather WeatherData
	if err = json.NewDecoder(resp.Body).Decode(&weather); err != nil {
		fmt.Println(err)
		i.Error("Unable to load weather data.")
		return
	}

	units, unitsErr := data.GetOption("units")

	// Basic embed
	embed := types.Embed{
		URL: fmt.Sprintf("https://darksky.net/forecast/%f,%f/", weather.Location.Lat, weather.Location.Lon),
		Thumbnail: types.EmbedImage{
			URL: "https:" + weather.Current.Condition.Icon,
		},
		Footer: types.EmbedFooter{
			Text: fmt.Sprintf("Conditions in %s, %s", weather.Location.Name, weather.Location.Region),
		},
		Fields: []types.EmbedField{
			{
				Name:   "☀️ Sunrise",
				Value:  weather.Forecast.Forecastday[0].Astro.Sunrise,
				Inline: true,
			},
			{
				Name:   fmt.Sprintf(":%s: Sunset", moonPhaseEmoji(weather.Forecast.Forecastday[0].Astro.MoonPhase)),
				Value:  weather.Forecast.Forecastday[0].Astro.Sunset,
				Inline: true,
			},
			{
				Name:   "\u200B",
				Value:  "\u200B",
				Inline: true,
			},
			{
				Name:   "Wind",
				Value:  "",
				Inline: true,
			},
			{
				Name:   "Humidity",
				Value:  fmt.Sprintf("%d%%", weather.Current.Humidity),
				Inline: true,
			},
			{
				Name:   "Visibility",
				Value:  "",
				Inline: true,
			},
		},
	}

	// Shortcuts for rain and snow data
	rainChance := weather.Forecast.Forecastday[0].Day.DailyChanceOfRain
	snowChance := weather.Forecast.Forecastday[0].Day.DailyChanceOfSnow

	// Add rain/snow warnings as appropriate
	if rainChance > 0 {
		embed.Description = fmt.Sprintf("There is a %d%% chance of rain", rainChance)
		if snowChance > 0 {
			embed.Description += fmt.Sprintf(" and a %d%% chance of snow.", snowChance)
		} else {
			embed.Description += "."
		}
	} else if snowChance > 0 {
		embed.Description = fmt.Sprintf("There is a %d%% chance of snow.", snowChance)
	}

	// Add imperial/metric data where appropriate
	if (unitsErr == nil && units.StringValue() == "metric") || weather.Location.Country != "United States of America" {
		embed.Title = fmt.Sprintf("%.1f°C, %s", weather.Current.TempC, weather.Current.Condition.Text)
		embed.Fields[3].Value = fmt.Sprintf("%.1f kph %s", weather.Current.WindKph, weather.Current.WindDir)
		embed.Fields[5].Value = fmt.Sprintf("%.1f km", weather.Current.VisKm)
	} else {
		embed.Title = fmt.Sprintf("%.1f°F, %s", weather.Current.TempF, weather.Current.Condition.Text)
		embed.Fields[3].Value = fmt.Sprintf("%.1f mph %s", weather.Current.WindMph, weather.Current.WindDir)
		embed.Fields[5].Value = fmt.Sprintf("%.1f mi", weather.Current.VisMiles)
	}

	// Respond with weather data
	i.Respond(types.InteractionResponse{
		Type: types.ResponseChannelMessageWithSource,
		Data: types.ResponseData{
			Embeds: []*types.Embed{&embed},
		},
	})

}

// Format sunset emoji with the right moon phase
func moonPhaseEmoji(phase string) string {
	phase = strings.ToLower(phase)
	if strings.Contains(phase, "moon") {
		return strings.ReplaceAll(phase, " ", "_")
	} else {
		return strings.ReplaceAll(phase+" moon", " ", "_")
	}
}
